jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,460 @@
+        {
+            if (LOGGER.isDebugEnabled())
+                LOGGER.debug("Attempt to remove an object from the pool that does not exist: {}", entry);
+        }
+
+        return removed;
+    }
+
+    public int size()
+    {
+        return sharedList.size();
+    }
+
+    public Collection<Entry> values()
+    {
+        return Collections.unmodifiableCollection(sharedList);
+    }
+
+    @Override
+    public void dump(Appendable out, String indent) throws IOException
+    {
+        Dumpable.dumpObjects(out, indent, this);
+    }
+
+    @Override
+    public String toString()
+    {
+        return getClass().getSimpleName() + " size=" + sharedList.size() + " closed=" + closed + " entries=" + sharedList;
+    }
+
+    public class Entry
+    {
+        // hi: positive=open/maxUsage counter,negative=closed lo: multiplexing counter
+        private final AtomicBiInteger state;
+        private volatile T pooled;
+
+        public Entry()
+        {
+            this.state = new AtomicBiInteger(-1, 0);
+        }
+
+        public T getPooled()
+        {
+            return pooled;
+        }
+
+        public void enable(T pooled)
+        {
+            if (!isClosed())
+                throw new IllegalStateException("Open entries cannot be enabled : " + this);
+            Objects.requireNonNull(pooled);
+            this.pooled = pooled;
+            state.set(0, 0);
+        }
+
+        /**
+         * @return true if the pooled object was acquired.
+         */
+        public boolean tryAcquire()
+        {
+            while (true)
+            {
+                long encoded = state.get();
+                int usageCount = AtomicBiInteger.getHi(encoded);
+                boolean closed = usageCount < 0;
+                int multiplexingCount = AtomicBiInteger.getLo(encoded);
+                int currentMaxUsageCount = maxUsageCount;
+                if (closed || multiplexingCount >= maxMultiplex || (currentMaxUsageCount > 0 && usageCount >= currentMaxUsageCount))
+                    return false;
+
+                if (state.compareAndSet(encoded, usageCount + 1, multiplexingCount + 1))
+                    return true;
+            }
+        }
+
+        /**
+         * @return false if the pooled object should be disposed of.
+         */
+        public boolean tryRelease()
+        {
+            int newMultiplexingCount;
+            int usageCount;
+            while (true)
+            {
+                long encoded = state.get();
+                usageCount = AtomicBiInteger.getHi(encoded);
+                boolean closed = usageCount < 0;
+                if (closed)
+                    return false;
+
+                newMultiplexingCount = AtomicBiInteger.getLo(encoded) - 1;
+                if (newMultiplexingCount < 0)
+                {
+                    if (LOGGER.isDebugEnabled())
+                        LOGGER.debug("try release failed on negative multiplexing count : {}", newMultiplexingCount);
```

Comment:
I'd rather throw `IllegalStateException` here - if we get this wrong we are going to screw anyway somewhere else.